Kadai1 dnakano

dai65527 opened the pull request:
課題1の画像変換コマンドを作成しました。お時間ある方レビューをお願いいたします。

- `convert.go` : main関数
- `imageConverter/imageConverter.go`: imageConverter モジュール

# 機能
コマンドライン引数に指定したディレクトリ以下の画像ファイルを変換します。

## 対応ファイル形式

- jpg -> png
- png -> jpg

※ その他の形式のファイル/ディレクトリは無視されます。

# 使い方
## コンパイル

```
go build convert.go
```

## 実行
jpgからpng

```
./convert directoryname -i=jpg -o=png
```

pngからjpg

```
./convert directoryname -i=png -o=jpg
```

## オプション

`-i`: 入力画像形式（デフォルト: png）
`-o`: 出力画像形式（デフォルト: jpg）

## テスト用ファイルダウンロードスクリプト
https://golang.org/doc/gopher よりGopherくんの画像をimagesディレクトリ内に格納します。

```
./getimages.sh
```

dai65527 committed:
add: main which can convert single jpg to png

dai65527 committed:
add imageConverter package

dai65527 committed:
Mod to use module

dai65527 committed:
rename main.go to convert.go

dai65527 committed:
mod: delete recusively calling convert function

dai65527 committed:
add: getimages.sh

script to get images

dai65527 committed:
add gitignore

dai65527 committed:
Add: imageConverter.Png2jpg

dai65527 committed:
Add comment to go doc

dai65527 committed:
improve error check of argument

- add existance
- set exit code

dai65527 committed:
add flag `-i` and `-o`

dai65527 committed:
add README.md

gosagawa commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,77 @@
+		fmt.Fprintln(os.Stderr, err)
+		return
+	}
+	fs.Close()
```
fs.Close()は呼ばない限りファイルを開いた状態のままになってしまうので必ず呼ぶ必要があるのですが、途中でreturnされると呼ばれなくなってしまうのでそれを避けるためにos.Openとセットでdeferを使うのが一般的かつ間違いがないかと思います。
```
	fs, err := os.Open(filename)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return
	}
        defer fs.Close()
```

またそうしたときに処理の流れとして以下の順番で呼ばれる事になるので最初のfsがクローズされない事になります。変数名を変えるか、別関数に切り分ける方が良さそうです。
```
fs, err := os.Open(filename)
fs, err = os.OpenFile(filename+".png", os.O_CREATE|os.O_WRONLY, 0644)
fs.Close()
fs.Close()
```

dai65527 replied:
なるほど、deferを使えばClose忘れをほぼ確実に防げるわけですね。ただ、上書きされてしまうとまずいで注意しないといけないと。とても参考になります。

gosagawa commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,77 @@
+
+	fs, err := os.Open(filename)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
```
ここでos.Stderrに出力してしまう事でこの関数がcliでの動作前提となってしまうので（もちろん課題としてはcliでしか動かさないのですが）、関数としてerrorを返し、main.goの方で標準エラーを出力した方が良さそうです。

dai65527 replied:
パッケージとして考えるならその方が確実にいいですね。修正することにしました！

gosagawa commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+			}
+		}
+		return nil
+	})
```
imageConverter.goのL22の指摘ともつながりますが、これだとerrは必ずnilになるので、ここ以下のコードは通らなそうです。Png2jpgおよびJpg2pngでエラーを返し、エラーがあったらそれを返すようにする方が良さそうです。

dai65527 replied:
filePath.Walk自体が失敗する可能性も考えてこうしてみました。画像変換コマンドからエラーが返ってくる場合も考慮してみます！

gosagawa commented on kadai1/getimages.sh:
```diff
@@ -0,0 +1,24 @@
+# reset and create nested images file
```
テストデータ準備用のシェルがあるのは良いですね！

gosagawa commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+// Jpg2png takes path to image of jpg and convert to png
+func Jpg2png(filename string) {
```
今回はjpgとpngのみなのでこれで大丈夫ですが、gifが追加になったら同じ作りだと煩雑になると思います。変換する形式を受け取りつつ汎用的な関数にしてgifを対応してみるのも余力があれば良いトレーニングになるのではと思います。

dai65527 replied:
そうですね。引数に拡張子を受け取る汎用的な関数に変更しました。（しかしgifには未対応...）

dai65527 committed:
modify imageConverter

- create new function Convert which can convert both jpg -> png, png -> jpg
- change not to print error in case of error but return error
- use defer to close file

dai65527 committed:
change to use imageConverter.Convert

dai65527 committed:
fix: add shebang

dai65527 committed:
mod typo

tenntenn commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+	"os"
+	"path/filepath"
+
+	"github.com/dai65527/gopherdojo-studyroom/kadai1/imageConverter"
```
パッケージ名はすべて小文字

dai65527 replied:
加えて短くシンプルな名前にする方がパッケージ使用者にとってもよい、ということですね！

tenntenn commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+	if len(flag.Args()) != 1 {
+		fmt.Fprintln(os.Stderr, "error: invalid argument")
+		os.Exit(1)
+	} else if _, err := os.Stat(flag.Args()[0]); err != nil {
```
elseいらない

dai65527 replied:
迷っていたところでした！ 前のブロックでreturn or exitする場合else不要。前段のifの条件を後ろに引き継がないようがよい。

tenntenn commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	// check flag is valid
+	if *flgInput != "jpg" && *flgInput != "jpeg" && *flgInput != "png" {
```
swithを使ってdefault句でやるとわかりやすくかける

```go
swtich *flgInput {
case "jpg", "jpeg", "png":
    // noop
default:
   // ここに書く
}
```

dai65527 replied:
可読性と保守性大幅に上がりますね。今後参考にします！

tenntenn commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+		if !info.IsDir() && filepath.Ext(path) == "."+inputFileExt {
+			erri := imageConverter.Convert(path, imageConverter.Extension(inputFileExt), imageConverter.Extension(outputFileExt))
+			if erri != nil {
+				fmt.Fprintln(os.Stdout, "error: "+path+": ", err)
```
main関数以外では基本的にerrorを返す

dai65527 replied:
エラー処理の考え方参考になります。

tenntenn commented on kadai1/convert.go:
```diff
@@ -0,0 +1,53 @@
+
+func convert(dirpath string, inputFileExt string, outputFileExt string) {
+	err := filepath.Walk(dirpath, func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() && filepath.Ext(path) == "."+inputFileExt {
```
```go
if info.IsDir() {
     return nil
}
```

dai65527 replied:
潜る場合はなるべく早く浮き上がる、意識します！

tenntenn commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,72 @@
+type Extension string
+
+// Convert takes a filename of image and convert it to another imagefile of outputFileExt
+func Convert(filename string, inputFileExt Extension, outputFileExt Extension) error {
```
`inputFileExt, outputFileExt Extension`とかける。
型で拡張子ということがわかってるので`in, out Extension`と書くほうがわかりやすい

dai65527 replied:
型が説明的だとすごくシンプルにわかりやすくなるのですね。ありがとうございます。

tenntenn commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,72 @@
+	outputFileExt = Extension(strings.ToLower(string(outputFileExt)))
+
+	// check extension
+	if inputFileExt != "jpg" && inputFileExt != "jpeg" && inputFileExt != "png" {
```
switchにするとよさそう

tenntenn commented on kadai1/imageConverter/imageConverter.go:
```diff
@@ -0,0 +1,72 @@
+	if err != nil {
+		return errors.New("failed to open file")
+	}
+	defer fs.Close()
```
書き込みなのでエラーチェックする

dai65527 replied:
勉強になりました。書き込みの際は正常に書き込まれてない場合を確認するようにします。

dai65527 committed:
change package name

imageConverter to imgcnv

dai65527 committed:
Recator

- use switch statement instead of if to check extension
- rename variables

dai65527 committed:
add error check when closing written file
